### PR TITLE
[Draft] Fix lax.max grad

### PR DESCRIPTION
**Pull request by joaogui1**
This solves both #2140 and #2107 

Here's the grad of softplus, using @shoyer 's code in #2107. 
image

Edit: Clarifying my PR, previously JAX considered `grad max(x,y) = 0.5`, when `x = y`. 
By using this [definition of max](https://math.stackexchange.com/questions/1237239/what-is-the-derivative-of-max-and-min-functions) and JAX's `grad abs(x - y) = 1`, when `x == y`, it makes sense for `grad max(x, y) = 1`, when `x == y`. 
Also by making the change we can solve both issues cited above

### CODE REVIEW

**Comment by joaogui1** [comment]
The Hessian test was using the fact that the grad of max(x, x) = 0.5, since I'm proposing setting that grad to 1.0 I had to change the test

**Comment by fehiepsi** [comment]
`grad max(x,y) = 0.5` makes sense to me. At least, we won't face grad issue of `x = max(x, 0) + min(x, 0)`. :D I found the definition helpful for [cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits), where the function is `max(x, 0) - x * z + log(1 + exp(-abs(x)))`, which involves both `max` and `abs`. With `grad max(x,y) = 0.5`, the gradient of `cross_entropy` will be correct.

**Comment by joaogui1** [comment]
If we use let the grad of min(0, 0) = 0.0 the first part is a non-issue I believe, but indeed getting the correct grad for cross-entropy is extremely important. 
Bummer, I wonder if there's a definition of the grad that works for all the cases.
In any case, the original problem this PR was meant to fix is going to be fixed by a custom grad and as you pointed out I would end up breaking other stuff, so I'm closing this, thanks for the input!

**Comment by yingted** [comment]
I think there's a definition of grad that's consistent with the C/NumPy functions, if we ignore NaN:
max(x,y) = the larger one, or x if they're equal
abs(x) = x if x>0 or x is +0, otherwise -x
Then, max(-0,0) = -0, max(0,-1) = 0, and abs(-0) = 0.

The downside is that it probably breaks even more things.
You can still do many things, but not as easily:
- (min(x,y), max(x,y)) becomes (min(x,y), max(y,x))  (stable sort)
- max(x,0) + min(x,0) becomes max(x,0) + min(0,x)  (same as above, y=0)
- (min(x,y), abs(x-y)) becomes (min(x,y), max(y,x)-min(x,y))
